File: common/src/main/java/com/zk/common/serial/SerialHelper.java
package com.zk.common.serial;

import com.zk.common.utils.LogUtil;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.List;

import android_serialport_api.SerialPort;

/**
 * Created by ZK on 2018/1/2.
 * 串口辅助工具类
 */
public abstract class SerialHelper {
    private SerialPort mSerialPort;
    private OutputStream mOutputStream;
    private InputStream mInputStream;
    private ReadThread mReadThread;
    private SendThread mSendThread;

    private String sPort = "/dev/ttyHSL1";
    private int iBaudRate = 115200;

    private boolean isOpen = false;
    private boolean sendFlag = true;

    private List<byte[]> sendList;
    private Object addSendSyncLock;

    public void open() throws SecurityException, IOException, InvalidParameterException {
        if (!isOpen) {

            addSendSyncLock = new Object();
            sendList = new ArrayList<>();

            isOpen = true;

            mSerialPort = new SerialPort(new File(sPort), iBaudRate, 0);

            mOutputStream = mSerialPort.getOutputStream();
            mInputStream = mSerialPort.getInputStream();

            mReadThread = new ReadThread();
            mReadThread.start();

            mSendThread = new SendThread();
            mSendThread.start();
        }

    }

    /**
     * 1、结束读取线程
     * 2、关闭SerialPort
     */
    public void close() {
        if (mReadThread != null)
            mReadThread.interrupt();
        if (mSerialPort != null) {
            mSerialPort.close();
            mSerialPort = null;
        }
        isOpen = false;
    }

    /**
     * 结束写的线程
     */
    public void stopSend() {
        if (mSendThread != null)
            mSendThread.interrupt();
        sendFlag = false;
        if (addSendSyncLock != null) {
            synchronized (addSendSyncLock) {
                addSendSyncLock.notify();
            }
        }
    }

    /**
     * 写
     *
     * @param bOutArray
     */
    private void send(byte[] bOutArray) {
        StringBuilder buffers = new StringBuilder();
        for (int i = 0; i < bOutArray.length; i++) {
            buffers.append(Integer.toHexString((bOutArray[i] & 0xff)));
            buffers.append(" ");
        }
        LogUtil.Companion.getInstance().d("SerialHelper", "test 发送" + buffers);

        try {
            mOutputStream.write(bOutArray);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * 读取线程
     */
    private class ReadThread extends Thread {
        @Override
        public void run() {
            super.run();
            while (!isInterrupted()) {
                try {
                    if (mInputStream == null) return;
                    byte[] buffer = new byte[4096];
                    int size = mInputStream.read(buffer);
                    if (size > 0) {
                        byte[] readBuffer = new byte[size];
                        System.arraycopy(buffer, 0, readBuffer, 0, size);
                        onDataReceived(sPort, readBuffer, size);
                    }
                } catch (Throwable e) {
                    e.printStackTrace();
                    return;
                }
            }
        }
    }

    /**
     * 写的线程
     */
    private class SendThread extends Thread {
        @Override
        public void run() {
            super.run();
            while (!isInterrupted()) {
                while (sendFlag) {
                    if (sendList == null || sendList.size() == 0) {
                        try {
                            synchronized (addSendSyncLock) {
                                addSendSyncLock.wait();
                            }
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    } else {
                        byte[] send = sendList.remove(0);
                        if (send != null) {
                            LogUtil.Companion.getInstance().d("串口通信Android发送：", send, send.length);

                            send(send);
                        }
                        try {
                            SendThread.sleep(200);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                }

            }
        }
    }

    /**
     * 添加写入的队列
     *
     * @param sendDate
     */
    public void addSendTask(byte[] sendDate) {
        sendList.add(sendDate);
        if (sendList.size() >= 1)
            synchronized (addSendSyncLock) {
                addSendSyncLock.notify();
            }
    }

    /**
     * 获取频率
     *
     * @return
     */
    public int getBaudRate() {
        return iBaudRate;
    }

    /**
     * 设置频率
     *
     * @param iBaud
     * @return
     */
    public boolean setBaudRate(int iBaud) {
        if (isOpen) {
            return false;
        } else {
            iBaudRate = iBaud;
            return true;
        }
    }

    /**
     * 设置频率
     *
     * @param sBaud
     * @return
     */
    public boolean setBaudRate(String sBaud) {
        int iBaud = Integer.parseInt(sBaud);
        return setBaudRate(iBaud);
    }

    /**
     * 获取串口路径
     *
     * @return
     */
    public String getPort() {
        return sPort;
    }

    /**
     * 设置串口路径
     *
     * @param sPort
     * @return
     */
    public boolean setPort(String sPort) {
        if (isOpen) {
            return false;
        } else {
            this.sPort = sPort;
            return true;
        }
    }

    /**
     * 是否打开读写线程
     *
     * @return
     */
    public boolean isOpen() {
        return isOpen;
    }

    /**
     * 收到的数据回调
     *
     * @param sPort  串口端口
     * @param buffer 数据
     * @param size   长度
     */
    protected abstract void onDataReceived(String sPort, byte[] buffer, int size);
}

File: common/src/main/java/com/zk/common/serial/SerialHelper.kt
package com.zk.common.serial

import com.zk.common.serial.android_serialport_api.SerialPort
import com.zk.common.utils.LogUtil
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.security.InvalidParameterException
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Created by ZK on 2018/1/2.
 * 串口辅助工具类
 */
abstract class SerialHelper(port: String, baudRate: Int) {
    private lateinit var mSerialPort: SerialPort
    private lateinit var mOutputStream: OutputStream
    private lateinit var mInputStream: InputStream
    private lateinit var mReadThread: ReadThread
    private var mSendThread: SendThread? = null

    private var mSendFlag = true

    // 待发送数据队列
    private val mSendList = LinkedList<ByteArray>() //TODO LinkedList理解的不是很深入
    // 写线程锁
    private val mSendLock = ReentrantLock()
    private val mSendCondition = mSendLock.newCondition()

    private var isOpen = false
    val mPort = port
    val mBaudRate = baudRate


    @Throws(SecurityException::class, IOException::class, InvalidParameterException::class)
    open fun open(withSend: Boolean) {
        if (!isOpen) {
            isOpen = true
            mSendFlag = true

            mSerialPort = SerialPort(File(mPort), mBaudRate, 0)
            mOutputStream = mSerialPort.outputStream
            mInputStream = mSerialPort.inputStream

            mReadThread = ReadThread()
            mReadThread.start()

            if (withSend) {
                mSendThread = SendThread()
                mSendThread!!.start()
            }
        }
    }

    open fun close() {
        mReadThread.interrupt()
        mSerialPort.close()
        isOpen = false
        if (mSendThread != null) mSendThread!!.interrupt()
        mSendFlag = false
        mSendLock.withLock {
            mSendCondition.signal()
        }
    }

    open fun addSendTask(sendDate: ByteArray): Boolean {
        var ret = false
        if (mSendThread != null) {
            mSendLock.withLock {
                ret = mSendList.offer(sendDate)
                mSendCondition.signal()
            }
        }
        return ret
    }

    /**
     * 读取线程
     */
    private inner class ReadThread : Thread() {
        override fun run() {
            super.run()
            while (!isInterrupted) {
                try {
                    val buffer = ByteArray(2048)
                    val size: Int = mInputStream.read(buffer)
                    if (size > 0) {
                        val readBuffer = ByteArray(size)
                        System.arraycopy(buffer, 0, readBuffer, 0, size)
                        onDataReceived(mPort, readBuffer, size)
                    }
                } catch (e: Throwable) {
                    e.printStackTrace()
                    return
                }
            }
        }
    }

    private inner class SendThread : Thread() {
        override fun run() {
            super.run()
            while (!isInterrupted) {
                while (mSendFlag) {
                    mSendLock.withLock {
                        if (mSendList.size == 0) {
                            LogUtil.instance.d("SerialHelper", "mSendCondition.await", false)
                            mSendCondition.await(1, TimeUnit.SECONDS)
                        }
                    }
                    if (mSendList.size > 0) {
                        var dataOperating: ByteArray? = null
                        try {
                            dataOperating = mSendList.poll()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                        if (dataOperating != null) {
                            // 写数据
                            send(dataOperating)
                        }
                        //非常重要，每发完一针一定要sleep
                        try {
                            sleep(200)
                        } catch (e: InterruptedException) {
                            e.printStackTrace()
                        }
                    }
                }
            }
        }
    }

    private fun send(outArray: ByteArray) {
        LogUtil.instance.d("串口通信Android发送：", outArray, outArray.size)
        try {
            mOutputStream.write(outArray)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    /**
     * 是否打开读写线程
     */
    fun isOpen(): Boolean {
        return isOpen
    }

    /**
     * 收到的数据回调
     *
     * @param sPort  串口端口
     * @param buffer 数据
     * @param size   长度
     */
    protected abstract fun onDataReceived(sPort: String?, buffer: ByteArray, size: Int)
}